.github/workflows/ci.yml

```diff
@@ -0,0 +1,265 @@
+      uses: actions-rs/toolchain@v1
+      with:
+        toolchain: ${{ matrix.rust }}
+        override: true
+
+    - name: Cache cargo state
+      uses: actions/cache@v2
+      env:
+        cache-name: cache-cargo
+      with:
+        path: |
+          ~/.cargo/registry
+          ~/.cargo/git
+        key: ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-${{ hashFiles('**/Cargo.toml') }}
+        restore-keys: |
+          ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-
+          ${{ matrix.os }}-${{ matrix.rust }}-build-
+          ${{ matrix.os }}-${{ matrix.rust }}-
+          ${{ matrix.os }}-
+
+    - name: Cache target dir
+      uses: actions/cache@v2
+      env:
+        cache-name: cache-target
+      with:
+        path: |
+          target/
+        key: ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-${{ hashFiles('**/Cargo.toml') }}
+        restore-keys: |
+          ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-
+          ${{ matrix.os }}-${{ matrix.rust }}-build-
+          ${{ matrix.os }}-${{ matrix.rust }}-
+          ${{ matrix.os }}-
+
+    - name: Build
+      uses: actions-rs/cargo@v1
+      with:
+        command: build
+        args: --all --release
+
+    - name: Test
+      uses: actions-rs/cargo@v1
+      with:
+        command: test
+        args: --all --all-features --release
+
+  clippy:
+    needs: build-and-test
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: nightly
+          override: true
+          components: clippy
+
+      - name: Cache cargo state
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-cargo
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-${{ hashFiles('**/Cargo.toml') }}
+          restore-keys: |
+            ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-
+            ${{ matrix.os }}-${{ matrix.rust }}-build-
+            ${{ matrix.os }}-${{ matrix.rust }}-
+            ${{ matrix.os }}-
+
+      - name: Cache target dir
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-target
+        with:
+          path: |
+            target/
+          key: ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-${{ hashFiles('**/Cargo.toml') }}
+          restore-keys: |
+            ${{ matrix.os }}-${{ matrix.rust }}-build-${{ env.cache-name }}-
+            ${{ matrix.os }}-${{ matrix.rust }}-build-
+            ${{ matrix.os }}-${{ matrix.rust }}-
+            ${{ matrix.os }}-
+
+      - uses: actions-rs/clippy-check@v1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          args: --all-targets --all-features -- -D warnings
+
+  format:
+    needs: build-and-test
```

Comment:
Is this really necessary ? I'd rather have this information as soon as possible.

Reply:
Right now it first does a build. If that fails it gets reported immediately. To run the tests, you would first have to build  the individual crates anyway. A syntactic validation of testcases might take a littler longer, but in all other cases I think it shouldn't make a big difference time-wise. One downside is that the reporting is not as straight forward as it used to be (I think you have to click on the check to see which command went wrong).

Reply:
Just to be sure, I was speaking about the `format` check. It doesn't technically needs build and test to perform. And I'm wondering if having to wait for them both just to discover that there is a formatting issue is wise or not.

Reply:
Oh, sorry that was my mistake, I should get a coffee. I guess it is always a tradeoff: Spinning up machines takes time and prevents other jobs to run in the meantime. But I guess format runs fast enough to put it towards the front 👍 .

Reply:
All good, it runs in seconds when all the rest run in more than 15mins, I think it's fine to have it run first